app2/ota_client.py

```diff
@@ -370,117 +375,103 @@ def _create_symbolic_links(self, list_file, standby_path):
             slink.symlink_to(slinkf.srcpath)
             os.chown(slink, slinkf.uid, slinkf.gid, follow_symlinks=False)
 
-    def _create_regular_files(self, url_base, cookies, list_file, standby_path):
+    def _create_regular_files(self, url_base: str, cookies, list_file, standby_path):
         lines = open(list_file).read().splitlines()
         reginf_list = [RegularInf(l) for l in lines]
         self._total_regular_files = len(reginf_list)  # via setter
+
         with Manager() as manager:
-            hardlink_dict = manager.dict()
             error_queue = manager.Queue()
             # NOTE: manager.Value doesn't work properly.
             processed_list = manager.list()
 
             def error_callback(e):
                 error_queue.put(e)
 
+            boot_standby_path = self._ota_status.get_standby_boot_partition_path()
+            _create_regfile_func = partial(
+                self._create_regular_file,
+                url_base=url_base,
+                cookies=cookies,
+                standby_path=standby_path,
+                processed_list=processed_list,
+                boot_standby_path=boot_standby_path,
+            )
+
             with Pool() as pool:
-                self._create_regular_files_with_async(
-                    url_base,
-                    cookies,
-                    standby_path,
-                    reginf_list,
-                    hardlink_dict,
-                    processed_list,
-                    pool.apply_async,
-                    error_callback,
-                )
+                hardlink_dict = dict()  # sha256hash[tuple[reginf, event]]
+
+                for reginf in reginf_list:
+                    if reginf.nlink >= 2:
+                        first_copy = not reginf.sha256hash in hardlink_dict
+                        prev_reginf, event = hardlink_dict.setdefault(
+                            reginf.sha256hash, (reginf, manager.Event())
+                        )
+
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf, prev_reginf),
+                            {"event": event, "first_copy": first_copy},
+                            error_callback=error_callback,
+                        )
+                    else:
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf,),
+                            error_callback=error_callback,
+                        )
+
                 pool.close()
                 while len(processed_list) < len(reginf_list):
-                    self._regular_files_processed = len(processed_list)  # via setter
```

Comment:
this line is required.

Reply:
Oh I miss understood the usage of this line, let me fix it.

Reply:
> to keep track of how many files are processed?

Yes!